### PR TITLE
Upgrade rails and other things

### PR DESCRIPTION
**Pull request by h-lame**
For: https://trello.com/c/vaecOj9d/493-upgrade-rails-on-licence-finder-2

Major points: upgrade rails to 4.2.7.1, ruby to 2.3.1, mongoid to 5.1.  On the way we upgrade as many other things that `bundle outdated` complained about as we can.  Interesting finding is that the `tire` gem we use to do searches has been retired by the author in favour of official gems released by elasticsearch.  We should do the work to migrate from tire to something else as it's a blocker to us upgrading beyond v. 33.1 of gds-api-adapters.  They both use rest-client, but tire is pinned to 1.x and newer gds-api-adapters wants 2.x.

### CODE REVIEW

**Comment by h-lame** [comment]
Flagging as DNM because there are no CI boxes that have both mongo 2.4 and elasticsearch 0.90 which is what our deployment infrastructure for this app is.  Loooking into getting this onto a newer elasticsearch which may require other changes.

**Comment by h-lame** [comment]
Removing DNM because we can finally get jenkins builds and such now that we've done the infrastructure changes (it's all on the trello card).

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@ def show_link_item(action, model, extra_params, &block)
     html << " class=\"#{item_class}\"".html_safe unless item_class.nil?
     html << ">\n".html_safe
     html << "<span class=\"#{key_name}-name\" id=\"#{model_id}\">".html_safe
-    html << "#{model.name}"
+    html << model.name.to_s
```
Comment (sihugh) [inline app/helpers/application_helper.rb]: We can probably remove the `to_s` altogether?

File: config/initializers/gds-api-adapters.rb [hunk 2]
```diff
@@ -0,0 +1 @@
+GdsApi.config.always_raise_for_not_found = true
```
Comment (davidbasalla) [inline config/initializers/gds-api-adapters.rb]: In other apps, we set the following as well:
```
  # Return a hash, not an OpenStruct from a request.
  config.hash_response_for_requests = true
```

Do we not need that in this app?

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (h-lame) [inline app/helpers/application_helper.rb]: Oh, yes, almost certainly.

File: config/initializers/gds-api-adapters.rb (hunk 2 above)
Comment (h-lame) [inline config/initializers/gds-api-adapters.rb]: Maybe we do. TBH, this PR is for general upgrade maintenance, separate from the story about the adapters.  There's work to be done to get it up to the most recent api adapters version too (see: https://github.com/alphagov/licence-finder/pull/82/commits/2cabcac0706c02c1392259a169b9fbb0cb418eb2).

Comment (h-lame) [inline config/initializers/gds-api-adapters.rb]: Actually, doesn't look like this is too onerous as it seems all the API interactions already uses hash accessors.  I'll do this too.

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (h-lame) [inline app/helpers/application_helper.rb]: done.

File: config/initializers/gds-api-adapters.rb (hunk 2 above)
Comment (h-lame) [inline config/initializers/gds-api-adapters.rb]: done.

Comment (davidbasalla) [inline config/initializers/gds-api-adapters.rb]: Ah nice 👍 True, had forgotten about that other story.

Comment (h-lame) [inline config/initializers/gds-api-adapters.rb]: Not hard, given how long this one has been on the go.

**Comment by davidbasalla** [comment]
Nice 👍 Merging...

**Pull request merged**